### PR TITLE
 Support Hibernate Subselect entity (fixes #519)

### LINKED ISSUE #519: Hibernate Subselect Entity not supported by EntityMetamodelImpl
### Description
After playing a bit with table functions (see also #181), I had the idea to experiment with mapping a `generate_series` query to an entity using the [`@Subselect`](https://docs.jboss.org/hibernate/orm/5.2/javadocs/org/hibernate/annotations/Subselect.html) annotation. Find below the produced stack trace.

A trivial work around is to wrap the query into to a view and map that instead. I am uncertain whether a fix for this is worthwhile, but it could perhaps be a step towards implementing table functions (#181).

```java
package com.pallasathenagroup.entities.series;

import lombok.Data;
import lombok.EqualsAndHashCode;
import org.hibernate.annotations.Immutable;
import org.hibernate.annotations.Subselect;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.Id;
import java.time.LocalDateTime;

@Data
@Entity
@Immutable
@Subselect("select start_date_time, start_date_time + interval '1 year' as end_date_time " +
    "from generate_series('2016-09-01'::timestamp, date_trunc('year', now()) + interval '8 months', '1 year') as start_date_time")
@EqualsAndHashCode(of = {"start"})
public class CollegeYearSeries implements Comparable<CollegeYearSeries> {

    @Id
    @Column(name = "start_date_time")
    private LocalDateTime start;

    @Column(name = "end_date_time")
    private LocalDateTime end;

    @Override
    public int compareTo(CollegeYearSeries o) {
        return getStart().compareTo(o.getStart());
    }
}
```

```
Caused by: java.lang.NullPointerException
	at com.blazebit.persistence.impl.hibernate.HibernateExtendedQuerySupport.getColumnTypes(HibernateExtendedQuerySupport.java:246)
	at com.blazebit.persistence.impl.EntityMetamodelImpl.<init>(EntityMetamodelImpl.java:110)
	at com.blazebit.persistence.impl.CriteriaBuilderFactoryImpl.<init>(CriteriaBuilderFactoryImpl.java:81)
	at com.blazebit.persistence.impl.CriteriaBuilderConfigurationImpl.createCriteriaBuilderFactory(CriteriaBuilderConfigurationImpl.java:599)
	at com.pallasathenagroup.iris.producers.CriteriaBuilderFactoryProducer.init(CriteriaBuilderFactoryProducer.java:32)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.as.ee.component.ManagedReferenceLifecycleMethodInterceptor.processInvocation(ManagedReferenceLifecycleMethodInterceptor.java:96)
	at org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:340)
	at org.jboss.as.weld.ejb.Jsr299BindingsInterceptor.doLifecycleInterception(Jsr299BindingsInterceptor.java:114)
	at org.jboss.as.weld.ejb.Jsr299BindingsInterceptor.processInvocation(Jsr299BindingsInterceptor.java:103)
	at org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:340)
	at org.jboss.invocation.InterceptorContext$Invocation.proceed(InterceptorContext.java:437)
	at org.jboss.weld.ejb.AbstractEJBRequestScopeActivationInterceptor.aroundInvoke(AbstractEJBRequestScopeActivationInterceptor.java:73)
	at org.jboss.as.weld.ejb.EjbRequestScopeActivationInterceptor.processInvocation(EjbRequestScopeActivationInterceptor.java:83)
	at org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:340)
	at org.jboss.as.weld.injection.WeldInjectionInterceptor.processInvocation(WeldInjectionInterceptor.java:53)
	at org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:340)
	at org.jboss.as.ee.component.ManagedReferenceFieldInjectionInterceptorFactory$ManagedReferenceFieldInjectionInterceptor.processInvocation(ManagedReferenceFieldInjectionInterceptorFactory.java:107)
	at org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:340)
	at org.jboss.as.ee.component.AroundConstructInterceptorFactory$1.processInvocation(AroundConstructInterceptorFactory.java:28)
	at org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:340)
	at org.jboss.as.weld.injection.WeldInterceptorInjectionInterceptor.processInvocation(WeldInterceptorInjectionInterceptor.java:56)
	at org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:340)
	at org.jboss.as.weld.ejb.Jsr299BindingsCreateInterceptor.processInvocation(Jsr299BindingsCreateInterceptor.java:100)
	at org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:340)
	at org.jboss.as.ee.component.NamespaceContextInterceptor.processInvocation(NamespaceContextInterceptor.java:50)
	at org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:340)
	at org.jboss.as.ejb3.tx.CMTTxInterceptor.invokeInOurTx(CMTTxInterceptor.java:275)
	... 29 more
```
### Environment

Version:            1.2.0.Alpha3
JPA-Provider:       Hibernate 5.2.12
DBMS:               PostgresSQL
Application Server: Wildfly

### PR DESCRIPTION
**Pull request by JWGmeligMeyling**
This is my attempt of fixing #519 . The reproducer testcase from #523 now passes, but I have a few questions regarding the implementation, and I haven't yet verified whether this approach will compile for all versions of Hibernate.

### CODE REVIEW

File: integration/hibernate-base/src/main/java/com/blazebit/persistence/impl/hibernate/HibernateJpaProvider.java [hunk 1]
```diff
@@ -502,12 +502,19 @@ protected String getMappedBy(CollectionPersister persister) {
         }
 
         // In this case, the property might represent a formula
-        if (columnNames.length == 1 && columnNames[0] == null) {
+        boolean isFormula = columnNames.length == 1 && columnNames[0] == null;
+        boolean isSubselect = entityPersister.getTableName().contains("(");
```
Comment (JWGmeligMeyling) [inline integration/hibernate-base/src/main/java/com/blazebit/persistence/impl/hibernate/HibernateJpaProvider.java]: Besides storing the SQL query segments in the entity persisters, the true value of the `Subselect` annotation only seems to be passed around during metamodel generation, and is not really obtainable through the metamodel. It may need some rework in order to support the *SQL from clause replacement * later on. An alternative that I have considered is `entityPersister.getEntityType().getReturnedClass().isAnnotationPresent(Subselect.class)`. Suggestions are really welcome here!

File: integration/hibernate-base/src/main/java/com/blazebit/persistence/impl/hibernate/HibernateJpaProvider.java [hunk 2]
```diff
@@ -502,12 +502,19 @@ protected String getMappedBy(CollectionPersister persister) {
...
+
+        if (isFormula || isSubselect) {
             Type propertyType = entityPersister.getPropertyType(attributeName);
             long length;
             int precision;
             int scale;
             try {
+                if (propertyType instanceof org.hibernate.type.EntityType) {
+                    propertyType = ((org.hibernate.type.EntityType) propertyType).getIdentifierOrUniqueKeyType(sfi);
```
Comment (JWGmeligMeyling) [inline integration/hibernate-base/src/main/java/com/blazebit/persistence/impl/hibernate/HibernateJpaProvider.java]: This was required to get the `PrimaryKeyJoinColumn` from the reproducer test case to work. I am pretty sure that the column type for an `EntityType` indeed should be the type of its identifying type, but I am uncertain whether this approach is always correct.

File: integration/hibernate-base/src/main/java/com/blazebit/persistence/impl/hibernate/HibernateJpaProvider.java (hunk 1 above)
Comment (JWGmeligMeyling) [reply integration/hibernate-base/src/main/java/com/blazebit/persistence/impl/hibernate/HibernateJpaProvider.java]: I've switched to `tables.length == 1 && tables[0] == null` because my previous assumption has been shown to be wrong by CI. I don't really like the approach of `entityPersister.getEntityType().getReturnedClass().isAnnotationPresent(Subselect.class)`, because it has all sorts of edge cases (such as a Subselect in an inheritance hierarchy - not that I am aware that this would be allowed).

File: core/testsuite/src/test/hibernate/com/blazebit/persistence/testsuite/Issue519Test.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+        @Generated(GenerationTime.INSERT)
+        @OneToOne(mappedBy = "a")
+        private B b;
+
+
+    }
+
+    @Entity
+    @Subselect("SELECT 5 as b, a.id AS aId FROM A")
```
Comment (beikov) [inline core/testsuite/src/test/hibernate/com/blazebit/persistence/testsuite/Issue519Test.java]: You are missing an alias for the from item `A` which is why I think the test is failing

Comment (JWGmeligMeyling) [reply core/testsuite/src/test/hibernate/com/blazebit/persistence/testsuite/Issue519Test.java]: A that makes sense. Postgresql is case insensitive to this as far as I know. I'll fix the test.
